vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage_test.go

```diff
@@ -0,0 +1,217 @@
+limitations under the License.
+*/
+
+package controllerfetcher
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	autoscalingapi "k8s.io/api/autoscaling/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+func getKey(key string) scaleCacheKey {
+	return scaleCacheKey{
+		namespace: "ns",
+		groupResource: schema.GroupResource{
+			Group:    "group",
+			Resource: "resource",
+		},
+		name: key,
+	}
+}
+
+func getScale() *autoscalingapi.Scale {
+	return &autoscalingapi.Scale{}
+}
+
+func TestControllerCache_InitiallyNotPresent(t *testing.T) {
+	c := newControllerCacheStorage(time.Second, time.Second, 10*time.Second)
+	key := getKey("foo")
+	present, _, _ := c.Get(key.namespace, key.groupResource, key.name)
+	assert.False(t, present)
+}
+
+func TestControllerCache_Refresh_NotExisting(t *testing.T) {
+	key := getKey("foo")
+	c := newControllerCacheStorage(time.Second, time.Second, 10*time.Second)
+	present, _, _ := c.Get(key.namespace, key.groupResource, key.name)
+	assert.False(t, present)
+
+	// Refreshing key that isn't in the cache doesn't insert it
+	c.Refresh(key.namespace, key.groupResource, key.name, getScale(), nil)
+	present, _, _ = c.Get(key.namespace, key.groupResource, key.name)
+	assert.False(t, present)
+}
+
+func TestControllerCache_Insert(t *testing.T) {
+	key := getKey("foo")
+	c := newControllerCacheStorage(time.Second, time.Second, 10*time.Second)
+	present, _, _ := c.Get(key.namespace, key.groupResource, key.name)
+	assert.False(t, present)
+
+	c.Insert(key.namespace, key.groupResource, key.name, getScale(), nil)
+	present, val, err := c.Get(key.namespace, key.groupResource, key.name)
+	assert.True(t, present)
+	assert.Equal(t, getScale(), val)
+	assert.Nil(t, err)
+}
+
+func TestControllerCache_InsertAndRefresh(t *testing.T) {
+	key := getKey("foo")
+	c := newControllerCacheStorage(time.Second, time.Second, 10*time.Second)
+	present, _, _ := c.Get(key.namespace, key.groupResource, key.name)
+	assert.False(t, present)
+
+	c.Insert(key.namespace, key.groupResource, key.name, getScale(), nil)
+	present, val, err := c.Get(key.namespace, key.groupResource, key.name)
+	assert.True(t, present)
+	assert.Equal(t, getScale(), val)
+	assert.Nil(t, err)
+
+	c.Refresh(key.namespace, key.groupResource, key.name, nil, fmt.Errorf("err"))
+	present, val, err = c.Get(key.namespace, key.groupResource, key.name)
+	assert.True(t, present)
+	assert.Nil(t, val)
+	assert.Errorf(t, err, "err")
+}
+
+func TestControllerCache_InsertNoOverwrite(t *testing.T) {
+	key := getKey("foo")
+	c := newControllerCacheStorage(time.Second, time.Second, 10*time.Second)
+	present, _, _ := c.Get(key.namespace, key.groupResource, key.name)
+	assert.False(t, present)
+
+	c.Insert(key.namespace, key.groupResource, key.name, getScale(), nil)
+	present, val, err := c.Get(key.namespace, key.groupResource, key.name)
+	assert.True(t, present)
+	assert.Equal(t, getScale(), val)
+	assert.Nil(t, err)
+
+	// Doesn't overwrite previous values
```

Comment:
I'm not sure why we want this behaviour.
IIUC the result from second Insert is fresher so it should overwrite the old value (maybe just not deleteAfter).
If there is a valid reason, please add a comment about it.

Reply:
I don't expect this to come up in how we use the cache so I think this won't affect VPAs performance. And I find this easier to understand.

Please let me know if I should change the behavior

Reply:
I don't have very strong feelings regarding the behaviour itself but if we simply don't care about whether it's one way of the other, maybe let's not have a test for this to basically keep it undefined?